feat(operator): enforce RFC 1035 validation for TrainJob name

juniemariam opened the pull request:
<!--  Thanks for sending a pull request! Here are some tips for you:
1. If this is your first time, check our contributor guidelines: https://www.kubeflow.org/docs/about/contributing
2. To know more about Kubeflow Trainer, check the developer guide:
    https://github.com/kubeflow/trainer/blob/master/CONTRIBUTING.md
3. If you want *faster* PR reviews, check how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
-->

**What this PR does / why we need it**:
- Adds validation to enforce RFC 1035 compliance for `TrainJob.metadata.name`
- Checks for:
  - DNS-1035 format (lowercase, alphanumeric, `-`, etc.)
  - Max length of 63 characters

### Why this is needed

This is a follow-up to #2734. Previously, invalid `TrainJob` names caused JobSet webhook rejections during reconciliation. This PR ensures invalid names are caught earlier at the TrainJob admission webhook level, improving clarity and user experience.

### Related Issues

Fixes #2735  
Related to #2621, #2734

### Tests

- Added unit tests for valid, uppercase, and too-long names
- Verified via `go test ./pkg/webhooks/...` and `make test`
- Unit tests, Integration tests passed
- E2E test for PyTorch runtime passed  
- E2E test for OpenMPI runtime failed due to timeout

kannon92 reviewed (commented):
Curious if you should just add this validation to the api?

you could easily do this via a CEL expression and avoid having to add it to the webhooks.

andreyvelich reviewed (commented):
/ok-to-test

andreyvelich commented:
/retest

juniemariam commented:
Thanks for the suggestion @kannon92 — that’s a great point! I followed the pattern from PRs like #2734 and #2722, where similar validations used webhooks. I'm still getting familiar with the project, so if CEL would be a better fit, I’d really appreciate any guidance and happy to update the PR if needed.

kannon92 commented:
> Thanks for the suggestion @kannon92 — that’s a great point! I followed the pattern from PRs like #2734, where similar validations used webhooks. I'm still getting familiar with the project, so if CEL would be a better fit, I’d really appreciate any guidance and happy to update the PR if needed.

Hmm. I’ll leave that to the maintainers. I like to follow similar patterns as the other code.

cc @tenzen-y @andreyvelich

astefanutti commented:
@juniemariam you can find some information about CEL validation rules in https://kubernetes.io/blog/2022/09/23/crd-validation-rules-beta/ and the documentation at
https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules

The idea is to specify a CEL rule directly in the Golang API using the `+kubebuilder:validation:XValidation:rule` marker.

There is also the format library that conveniently exposes the `dns1035Label` function accessible from CEL rules: https://kubernetes.io/docs/reference/using-api/cel/#kubernetes-format-library

juniemariam commented:
Update: Switched from webhook-based validation to CEL-based validation for TrainJob.metadata.name.
- Added CEL validation rules to enforce RFC 1035 compliance
- Removed webhook logic for name validation
- Regenerated CRDs with make manifests and committed changes

Verified with:
- kubectl apply (valid + invalid test cases)
- make test (unit tests pass)
- also ran integration and E2E test cases (passed)

Let me know if further changes are needed!
/cc @andreyvelich  @astefanutti

juniemariam commented on pkg/webhooks/trainjob_webhook_test.go:
```diff
@@ -45,19 +45,6 @@ func TestValidateCreate(t *testing.T) {
 			wantError:    nil,
 			wantWarnings: nil,
 		},
-		"invalid trainjob name with uppercase letters": {
```
@astefanutti Just to clarify—now that the validation is enforced in the CRD using CEL, do we still need to keep the Go/webhook logic and its related unit tests?

I moved the tests out of the webhook layer because the validation logic now lives in the CRD, so the webhook no longer performs this check.

Could you please advise whether it’s better to keep both the Go logic and tests (alongside the CEL rule and integration test), or just rely on CEL + integration tests?

astefanutti replied:
@juniemariam that makes sense. Yes an integration test could be useful.

astefanutti commented on pkg/apis/trainer/v1alpha1/trainjob_types.go:
```diff
@@ -34,6 +34,8 @@ const (
 // +kubebuilder:storageversion
 // +kubebuilder:printcolumn:name="State",type=string,JSONPath=`.status.conditions[-1:].type`
 // +kubebuilder:printcolumn:name="Age",type=date,JSONPath=`.metadata.creationTimestamp`
+// +kubebuilder:validation:XValidation:rule="self.metadata.name.matches('^[a-z]([-a-z0-9]*[a-z0-9])?$')", message="metadata.name must match RFC 1035 DNS label format"
```
@juniemariam thanks for the explanation. That looks good to me as it is.

astefanutti commented:
/ok-to-test
